### PR TITLE
[FEATURE] Crafting GUI

### PR DESCRIPTION
**Pull request by seailz**
[contributing]: https://github.com/Negative-Games/Framework/blob/main/CONTRIBUTING.md

## Pull Request Etiquette

 

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

 

### Changes

- [x] Internal code
- [x] Library interface (affecting end-user code)
- [x] Documentation
- [ ] Other: \_____ 

 

Closes Issue: #99 

## Description

Added a crafting GUI

## Documentation
```java
public class ExampleCraftingGUI extends CraftingGUI {
    public ExampleCraftingGUI() {
        super("&6&lExample Crafting GUI");


        ItemStack diamond = new ItemStack(Material.DIAMOND);
        ItemStack iron = new ItemStack(Material.IRON_INGOT);
        ItemStack gold = new ItemStack(Material.GOLD_INGOT);
        ItemStack bedrock = new ItemStack(Material.BEDROCK);


        // Set an item without a click event
        // The reason why this is a function is because you could do conditional items.
        setItem(0, player -> {
            if (player.hasPermission("myserver.admin")) {
                return bedrock;
            } else {
                return diamond;
            }
        });

        // This is what you would do if you do not want conditional items.
        setItem(0, player -> diamond);

        // Set an item with a click event
        setItemClickEvent(1, player -> diamond, (player, event) -> {
            player.sendMessage("&aYou clicked the diamond!");
            player.closeInventory();
        });

        // Add an item to the next available slot without a click event
        addItem(player -> iron);

        // Add an item to the next available slot with a click event
        addItemClickEvent(player -> gold, (player, event) -> {
            player.sendMessage("&aYou clicked the gold!");
            player.closeInventory();
        });
    }
}
```

### CODE REVIEW

**Comment by seailz** [comment]
ok that's enough attending to issues for today hehe

**Comment by joeecodes** [comment]
Won't be merged but will be reviewed.

File: src/main/java/games/negative/framework/gui/CraftingGUI.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+
+/**
+ * Crafting GUI
+ * @author Negative - modified by Seailz
```
Comment (joeecodes) [inline src/main/java/games/negative/framework/gui/CraftingGUI.java:29]: leave the original author tag as it is. you don't need to add "modified". 

author the code that you write

**Review by joeecodes — CHANGES_REQUESTED** [review]
That's all i've got for now. @NegativeDev any thing to add?

File: src/main/java/games/negative/framework/gui/holder/CraftingGUIHolder.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import org.bukkit.event.inventory.InventoryOpenEvent;
+import org.bukkit.event.inventory.InventoryType;
+import org.bukkit.inventory.Inventory;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.function.BiConsumer;
+
+/**
+ * Crafting GUI Holder
+ * @author Negative - modified by Seailz
```
Comment (joeecodes) [inline src/main/java/games/negative/framework/gui/holder/CraftingGUIHolder.java:22]: Author notes need to be done in all classes

File: src/main/java/games/negative/framework/gui/listener/GUIListener.java [hunk 3]
```diff
@@ -56,6 +57,11 @@ public void onInvClick(InventoryClickEvent event) {
             base.onClick(event);
         }
 
+        if (holder instanceof CraftingGUIHolder) {
+            CraftingGUIHolder base = (CraftingGUIHolder) holder;
```
Comment (joeecodes) [inline src/main/java/games/negative/framework/gui/listener/GUIListener.java:61]: Try not to use single word vars, makes it hard to know what it is when referencing it. 

Keep it relevant to what the object it is for

**Review by joeecodes — CHANGES_REQUESTED** [review]

File: src/main/java/games/negative/framework/gui/listener/GUIListener.java (hunk 3 above)
Comment (seailz) [reply src/main/java/games/negative/framework/gui/listener/GUIListener.java:61]: Literally the same code from above with a few things tweaked?

**Comment by joeecodes** [comment]
Are there any screenshots of this being used?

**Comment by seailz** [comment]
> Are there any screenshots of this being used? Just so I can see what it actually looks like in production.

Will send a few in a couple hours

**Comment by seailz** [comment]
> Are there any screenshots of this being used? Just so I can see what it actually looks like in production.

there's one on the issue relating to this pr, but that's not created with this framework

**Comment by seailz** [comment]
> Are there any screenshots of this being used? Just so I can see what it actually looks like in production.

there's one on the issue relating to this pr, but that's not created with this framework

**Comment by joeecodes** [comment]
> > Are there any screenshots of this being used? Just so I can see what it actually looks like in production.
> 
> there's one on the issue relating to this pr, but that's not created with this framework

is it a thing that needs to be addressed by negative games or not?

**Comment by seailz** [comment]
> > > Are there any screenshots of this being used? Just so I can see what it actually looks like in production.
> > 
> > 
> > there's one on the issue relating to this pr, but that's not created with this framework
> 
> is it a thing that needs to be addressed by negative games or not?

wdym

**Comment by joeecodes** [comment]
> > > > Are there any screenshots of this being used? Just so I can see what it actually looks like in production.
> > > 
> > > 
> > > there's one on the issue relating to this pr, but that's not created with this framework
> > 
> > 
> > is it a thing that needs to be addressed by negative games or not?
> 
> wdym

You said theres an issue with this PR. Is it something that needs to be fixed by us?

**Comment by joeecodes** [comment]
ahhh

**Pull request closed without merging**